src/transports/smart.c

```diff
@@ -460,11 +460,28 @@ static int ref_name_cmp(const void *a, const void *b)
 	return strcmp(ref_a->head.name, ref_b->head.name);
 }
 
-int git_transport_smart_certificate_check(git_transport *transport, git_cert *cert, int valid, const char *hostname)
+int git_transport_smart_certificate_check(git_transport *transport, git_cert *cert, int cert_error, const char *hostname)
 {
 	transport_smart *t = (transport_smart *)transport;
+	git_error_state state = {0};
+	int cberror, error;
+	int is_valid = (cert_error == GIT_OK);
+
+	if (t->certificate_check_cb == NULL)
+		return cert_error;
+
+	giterr_state_capture(&state, cert_error);
+	cberror = t->certificate_check_cb(cert, is_valid, hostname, t->message_cb_payload);
+	if (cberror == GIT_PASSTHROUGH) {
+		error = giterr_state_restore(&state);
+	} else {
+		error = cberror;
+		if (error < 0 && !giterr_last())
+			giterr_set(GITERR_NET, "user cancelled certificate check");
+	}
+	giterr_state_free(&state);
 
-	return t->certificate_check_cb(cert, valid, hostname, t->message_cb_payload);
+	return error;
 }
```

Comment:
All this code shouldn't go here. This function is a way for code external to the library to execute the certificate check callback, The error you want to insert here may not exist in the code implementing a transport. They may be catching an exception several languages apart, and there isn't going to be any error state to capture.

If we need common code inside the library to do this repeated code, that needs its own function.

Reply:
I really wanted to make that passthrough-dance generic across all transports, and having it here removes the need for an external transport implementation to handle this case on their side. I can easily move it in a private function, but IMHO that kinda defeat the purpose, since it will no longer be "externally usable". I also like the eat-your-own-dogfood aspect of it.

The usage pattern I had in mind was that the external transport would "check the certificate" while connecting (whatever that means), call this with cert_error set to `GIT_OK` for valid, `GIT_ECERTIFICATE` for invalid (that's the "non-existing error case" from your point above, which our SSH transport demonstrates), `anything else` for spurious unrelated error, and this function would automatically preserve the callback error (which might not even be aware that an external transport is in use) or restore the transport's original success/failure state on a passthrough.

I'm not sure I understand your point about exceptions being several languages apart : should it be safe for an higher-level (say Ruby) external transport to raise an exception and catch that around the (say `git_remote_download`) call that started the activity, without going through our code at all 😨 ?

As an API sidenote, I was actually wondering about the usage pattern of those functions : the external code *might* end up calling a NULL function pointer, and I don't think there's a way for external users to know they're about to SEGV (`git_transport_smart_credentials` shares the same problem).